src/java.base/share/classes/java/util/zip/InflaterInputStream.java

```diff
@@ -155,7 +155,12 @@ public int read(byte[] b, int off, int len) throws IOException {
                     reachEOF = true;
                     return -1;
                 }
-                if (inf.needsInput()) {
+                if (inf.needsInput() && !inf.hasPendingOutput()) {
```

Comment:
I think the approach looks okay and avoids attempting to fill and the unspecified EOFException that it throws when the input is at EOF.